# Sleep Equation

## модель 1 

спецификация: 
$$
sleep = \beta_0 + \beta_1 * totwrk + \beta_2 * male + u
$$

In [1]:
import numpy as np # linear algebra 
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
# stats models: regression fitting via formulas
import statsmodels.formula.api as smf
# stats models: regression fitting via matrices of regression design
import statsmodels.api as sm

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/sleep75.csv')
df

age  black  case  clerical  construc  educ  earns74  gdhlth  inlf  leis1  \
0     32      0     1  0.000000  0.000000    12        0       0     1   3529   
1     31      0     2  0.000000  0.000000    14     9500       1     1   2140   
2     44      0     3  0.000000  0.000000    17    42500       1     1   4595   
3     30      0     4  0.000000  0.000000    12    42500       1     1   3211   
4     64      0     5  0.000000  0.000000    14     2500       1     1   4052   
..   ...    ...   ...       ...       ...   ...      ...     ...   ...    ...   
701   45      0   702  0.182331  0.030075    12     5500       1     0   5069   
702   34      0   703  0.182331  0.030075    10     2500       0     0   5885   
703   37      0   704  0.182331  0.030075    12     3500       1     0   4719   
704   54      0   705  0.182331  0.030075    17    32500       1     0   5149   
705   30      0   706  0.182331  0.030075    16     6750       1     0   4747   

     ...  spwrk75  totwrk  union  worknrm  workscnd  exper  yngkid  yrsmarr  \
0    ...        0    3438      0     3438         0     14       0       13   
1    ...        0    5020      0     5020         0     11       0        0   
2    ...        1    2815      0     2815         0     21       0        0   
3    ...        1    3786      0     3786         0     12       0       12   
4    ...        1    2580      0     2580         0     44       0       33   
..   ...      ...     ...    ...      ...       ...    ...     ...      ...   
701  ...        1    2026      0     2026         0     27       0       18   
702  ...        0     675      1      465       210     18       0        4   
703  ...        1    1851      0     1851         0     19       0       17   
704  ...        1    1961      1     1481       480     31       0       22   
705  ...        0    2363      0     2363         0      8       1        9   

        hrwage  agesq  
0     7.070004   1024  
1     1.429999    961  
2    20.529997   1936  
3     9.619998    900  
4     2.750000   4096  
..         ...    ...  
701        NaN   2025  
702        NaN   1156  
703        NaN   1369  
704        NaN   2916  
705        NaN    900  

[706 rows x 34 columns]

In [4]:
# Зададим матрицы регрессионного дизайна
y = df['sleep'] # зависимая переменная
X = sm.add_constant(df[['totwrk', 'male']]) 
X.head(n=7)

const  totwrk  male
0    1.0    3438     1
1    1.0    5020     1
2    1.0    2815     1
3    1.0    3786     0
4    1.0    2580     1
5    1.0    1205     1
6    1.0    2113     1

In [5]:
sleep_eq3 = sm.OLS(y, X).fit()
# Коэфициенты модели с округлением
sleep_eq3.params.round(2)

const     3573.20
totwrk      -0.17
male        88.84
dtype: float64

- При увеличении количества рабочих минут в неделю на единицу, кол-во сна в неделю уменьшается на 0.17
- Мужчины спят в среднем на 88.84 минуты больше

## модель 2 

$$
sleep = \beta_0 + \beta_1 * totwrk + \beta_2 * male +...+ u
$$

In [8]:
df[['sleep', 'totwrk', 'male', 'smsa', 'age', 'south', 'yngkid', 'marr', 'union']].describe()

sleep       totwrk        male        smsa         age  \
count   706.000000   706.000000  706.000000  706.000000  706.000000   
mean   3266.355524  2122.920680    0.566572    0.399433   38.815864   
std     444.413448   947.470123    0.495900    0.490129   11.342637   
min     755.000000     0.000000    0.000000    0.000000   23.000000   
25%    3015.000000  1553.500000    0.000000    0.000000   29.000000   
50%    3270.500000  2288.000000    1.000000    0.000000   36.000000   
75%    3532.250000  2691.750000    1.000000    1.000000   48.000000   
max    4695.000000  6415.000000    1.000000    1.000000   65.000000   

            south      yngkid       marr       union  
count  706.000000  706.000000  706.00000  706.000000  
mean     0.184136    0.128895    0.82153    0.218130  
std      0.387870    0.335321    0.38318    0.413269  
min      0.000000    0.000000    0.00000    0.000000  
25%      0.000000    0.000000    1.00000    0.000000  
50%      0.000000    0.000000    1.00000    0.000000  
75%      0.000000    0.000000    1.00000    0.000000  
max      1.000000    1.000000    1.00000    1.000000

In [9]:
# Зададим матрицы регрессионного дизайна
y = df['sleep'] # зависимая переменная

X = sm.add_constant(df[['totwrk', 'male', 'smsa', 'age', 'south', 'yngkid', 'marr', 'union']])  
X.tail(n=7)

const  totwrk  male  smsa  age  south  yngkid  marr  union
699    1.0    2725     1     0   61      0       0     1      0
700    1.0    2388     0     1   32      1       0     0      0
701    1.0    2026     0     1   45      0       0     1      0
702    1.0     675     1     0   34      0       0     1      1
703    1.0    1851     0     0   37      0       0     1      0
704    1.0    1961     0     0   54      1       0     1      1
705    1.0    2363     0     0   30      1       1     1      0

In [10]:
sleep_eq3 = sm.OLS(y, X).fit()
# Коэфициенты модели с округлением
sleep_eq3.params.round(2)

const     3446.83
totwrk      -0.17
male        87.11
smsa       -54.19
age          2.71
south      102.27
yngkid     -13.05
marr        31.36
union       11.87
dtype: float64

- При увеличении количества рабочих минут в неделю на единицу, кол-во сна в неделю уменьшается на 0.17.
- Мужчины спят в среднем на 87.11 минуты больше.

## модель 3 

$$
sleep = \beta_0 + \beta_1 * hrwage + ...+ u
$$

In [12]:
df[['sleep', 'hrwage', 'smsa', 'totwrk', 'male', 'marr', 'age', 'south', 'yngkid']].describe()

sleep      hrwage        smsa       totwrk        male  \
count   706.000000  532.000000  706.000000   706.000000  706.000000   
mean   3266.355524    5.082839    0.399433  2122.920680    0.566572   
std     444.413448    3.704385    0.490129   947.470123    0.495900   
min     755.000000    0.350000    0.000000     0.000000    0.000000   
25%    3015.000000    2.890002    0.000000  1553.500000    0.000000   
50%    3270.500000    4.379999    0.000000  2288.000000    1.000000   
75%    3532.250000    6.210001    1.000000  2691.750000    1.000000   
max    4695.000000   35.509987    1.000000  6415.000000    1.000000   

            marr         age       south      yngkid  
count  706.00000  706.000000  706.000000  706.000000  
mean     0.82153   38.815864    0.184136    0.128895  
std      0.38318   11.342637    0.387870    0.335321  
min      0.00000   23.000000    0.000000    0.000000  
25%      1.00000   29.000000    0.000000    0.000000  
50%      1.00000   36.000000    0.000000    0.000000  
75%      1.00000   48.000000    0.000000    0.000000  
max      1.00000   65.000000    1.000000    1.000000

In [13]:
# Зададим матрицы регрессионного дизайна
y = df['sleep'] # зависимая переменная
# в исходом датасете есть переменная lhrwage=log(hrwage)
X = sm.add_constant(df[['lhrwage', 'smsa', 'totwrk', 'male', 'marr', 'age', 'south', 'yngkid']])  
X.tail(n=7)

const  lhrwage  smsa  totwrk  male  marr  age  south  yngkid
699    1.0      NaN     0    2725     1     1   61      0       0
700    1.0      NaN     1    2388     0     0   32      1       0
701    1.0      NaN     1    2026     0     1   45      0       0
702    1.0      NaN     0     675     1     1   34      0       0
703    1.0      NaN     0    1851     0     1   37      0       0
704    1.0      NaN     0    1961     0     1   54      1       0
705    1.0      NaN     0    2363     0     1   30      1       1

In [14]:
sleep_eq3 = sm.OLS(y, X, missing='drop').fit()
# Коэфициенты модели с округлением
sleep_eq3.params.round(2)

const      3440.19
lhrwage      -1.39
smsa        -36.96
totwrk       -0.16
male         36.87
marr         53.34
age           2.37
south        76.27
yngkid       47.92
dtype: float64

- Люди живущие в мегаполисе в среднем спят меньше на 36.96 минут.
- При изменении почасовой оплаты на 1% человек спит на 0.0139 минуты в неделю меньше.